### PR TITLE
PROCESSINGS: UpsertPostImpressionAsync

### LINKED ISSUE #414: PROCESSINGS: UpsertPostImpressionAsync
See https://github.com/hassanhabib/The-Standard/blob/885cb87313055dd8e52fd9e3f247af5a5a71396f/2.%20Services/2.2%20Processings/2.2%20Processings.md

### PR DESCRIPTION
**Pull request by Javlonbek-dev**
Closes #414

### CODE REVIEW

File: Taarafo.Core/Models/Processings/PostImpressions/Exceptions/NullPostImpressionProcessingException.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+using Xeptions;
+
+namespace Taarafo.Core.Models.Processings.PostImpressions.Exceptions
+{
+    public class NullPostImpressionProcessingException : Xeption
+    {
+        public NullPostImpressionProcessingException()
+            : base(message: "PostImpression is null.") { }
+
```
Comment (cjdutoit) [inline Taarafo.Core/Models/Processings/PostImpressions/Exceptions/NullPostImpressionProcessingException.cs:14]: Remove line

File: Taarafo.Core/Services/Processings/PostImpressions/IPostImpressionProcessingService.cs [hunk 2]
```diff
@@ -4,12 +4,15 @@
...
+using System.Threading.Tasks;
 using Taarafo.Core.Models.PostImpressions;
 
 namespace Taarafo.Core.Services.Processings.PostImpressions
 {
     public interface IPostImpressionProcessingService
     {
+        ValueTask<PostImpression> UpsertPostImpressionAsync(PostImpression postImpression);
         IQueryable<PostImpression> RetrieveAllPostImpressions();
+
```
Comment (cjdutoit) [inline Taarafo.Core/Services/Processings/PostImpressions/IPostImpressionProcessingService.cs:16]: Remove line

File: Taarafo.Core/Services/Processings/PostImpressions/PostImpressionProcessingService.Exceptions.cs [hunk 3]
```diff
@@ -14,8 +15,48 @@ namespace Taarafo.Core.Services.Processings.PostImpressions
...
+                throw CreateAndLogValidationException(nullPostImpressionProcessingException);
+            }
+            catch (InvalidPostImpressionProcessingException invalidPostImpressionProcessingException)
+            {
+                throw CreateAndLogValidationException(invalidPostImpressionProcessingException);
+            }
+            catch (PostImpressionValidationException postImpressionValidationException)
+            {
+                throw CreateAndLogDependencyValidationException(postImpressionValidationException);
+            }
```
Comment (cjdutoit) [inline Taarafo.Core/Services/Processings/PostImpressions/PostImpressionProcessingService.Exceptions.cs:38]: You don't need this catch block.  All the validation on PostImpression is thrown as an `InvalidPostImpressionProcessingException` then catagorised to `PostImpressionValidationException` with the `throw CreateAndLogValidationException(invalidPostImpressionProcessingException);`

File: Taarafo.Core/Services/Processings/PostImpressions/PostImpressionProcessingService.cs [hunk 4]
```diff
@@ -23,6 +26,31 @@ public partial class PostImpressionProcessingService : IPostImpressionProcessing
             this.loggingBroker = loggingBroker;
         }
 
+        public ValueTask<PostImpression> UpsertPostImpressionAsync(PostImpression postImpression) =>
+        TryCatch(async () =>
```
Comment (cjdutoit) [inline Taarafo.Core/Services/Processings/PostImpressions/PostImpressionProcessingService.cs:30]: Indent the TryCatch

File: Taarafo.Core/Services/Processings/PostImpressions/PostImpressionProcessingService.cs [hunk 5]
```diff
@@ -23,6 +26,31 @@ public partial class PostImpressionProcessingService : IPostImpressionProcessing
...
+        {
+            IQueryable<PostImpression> postImpressions =
+                this.postImpressionService.RetrieveAllPostImpressions();
+
+            return postImpressions.FirstOrDefault(SamePostImpressionAs(postImpression));
+        }
+
+        private static Expression<Func<PostImpression, bool>> SamePostImpressionAs(PostImpression postImpression) =>
+            retrievePostImpression => (retrievePostImpression.PostId == postImpression.PostId)
+                || (retrievePostImpression.ProfileId == postImpression.ProfileId);
```
Comment (cjdutoit) [inline Taarafo.Core/Services/Processings/PostImpressions/PostImpressionProcessingService.cs:52]: You need to macth on both.  Change `||` to `&&`

**Review by cjdutoit — CHANGES_REQUESTED** [review]
Some changes required

File: Taarafo.Core/Models/Processings/PostImpressions/Exceptions/NullPostImpressionProcessingException.cs (hunk 1 above)
Comment (Javlonbek-dev) [reply Taarafo.Core/Models/Processings/PostImpressions/Exceptions/NullPostImpressionProcessingException.cs:14]: Done

File: Taarafo.Core/Services/Processings/PostImpressions/PostImpressionProcessingService.Exceptions.cs (hunk 3 above)
Comment (Javlonbek-dev) [reply Taarafo.Core/Services/Processings/PostImpressions/PostImpressionProcessingService.Exceptions.cs:38]: Done

File: Taarafo.Core/Services/Processings/PostImpressions/PostImpressionProcessingService.cs (hunk 5 above)
Comment (Javlonbek-dev) [reply Taarafo.Core/Services/Processings/PostImpressions/PostImpressionProcessingService.cs:52]: Done

File: Taarafo.Core/Services/Processings/PostImpressions/PostImpressionProcessingService.cs (hunk 4 above)
Comment (Javlonbek-dev) [reply Taarafo.Core/Services/Processings/PostImpressions/PostImpressionProcessingService.cs:30]: Done

Comment (Javlonbek-dev) [reply Taarafo.Core/Services/Processings/PostImpressions/PostImpressionProcessingService.cs]: Done

File: Taarafo.Core/Services/Processings/PostImpressions/IPostImpressionProcessingService.cs (hunk 2 above)
Comment (Javlonbek-dev) [reply Taarafo.Core/Services/Processings/PostImpressions/IPostImpressionProcessingService.cs]: Done

**Review by cjdutoit — APPROVED** [review]

**Pull request merged**